cli/crates/typed-resolvers/src/codegen.rs

```diff
@@ -33,10 +33,11 @@ where
 
                 for field in fields {
                     let field_optional = !is_input_object
-                        && matches!(
+                        && (matches!(
                             field.r#type.kind,
                             TypeKind::Definition(Definition::Object(_) | Definition::Union(_))
-                        );
+                        ) || field.resolver_name.is_some()
+                            || field.has_arguments);
```

Comment:
Not the prettiest formatting, happy to push something nicer if that seems better.